implementation/src/main/java/io/smallrye/health/SmallRyeHealthReporter.java

```diff
@@ -99,57 +129,169 @@ public void reportHealth(OutputStream out, SmallRyeHealth health) {
+        return getHealthGroupsAsync().await().indefinitely();
+    }
 
-        List<HealthCheck> groupHealthChecks = new ArrayList<>();
+    @Experimental("Asynchronous Health Check procedures")
+    public Uni<SmallRyeHealth> getHealthAsync() {
+        return getHealthAsync(HEALTH, LIVENESS, READINESS);
+    }
+
+    @Experimental("Asynchronous Health Check procedures")
+    public Uni<SmallRyeHealth> getLivenessAsync() {
+        return getHealthAsync(LIVENESS);
+    }
+
+    @Experimental("Asynchronous Health Check procedures")
+    public Uni<SmallRyeHealth> getReadinessAsync() {
+        return getHealthAsync(READINESS);
+    }
+
+    @Experimental("Asynchronous Health Check procedures and Health Groups")
+    public Uni<SmallRyeHealth> getHealthGroupAsync(String groupName) {
+        Instance<HealthCheck> checks = allHealthChecks.select(HealthGroup.Literal.of(groupName));
+        Instance<AsyncHealthCheck> asyncChecks = allAsyncHealthChecks.select(HealthGroup.Literal.of(groupName));
+
+        return getHealthAsync(Collections.singletonList(checks), Collections.singletonList(asyncChecks));
+    }
+
+    @Experimental("Asynchronous Health Check procedures and Health Groups")
+    public Uni<SmallRyeHealth> getHealthGroupsAsync() {
+        List<HealthCheck> groupHealthChecks = getHealthGroupsChecks(HealthCheck.class);
+        List<AsyncHealthCheck> groupAsyncHealthChecks = getHealthGroupsChecks(AsyncHealthCheck.class);
+
+        return getHealthAsync(Collections.singletonList(groupHealthChecks), Collections.singletonList(groupAsyncHealthChecks));
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T> List<T> getHealthGroupsChecks(Class<T> checkClass) {
+        Iterator<Bean<?>> iterator = beanManager.getBeans(checkClass, Any.Literal.INSTANCE).iterator();
+
+        List<T> groupHealthChecks = new ArrayList<>();
 
         while (iterator.hasNext()) {
             Bean<?> bean = iterator.next();
             if (bean.getQualifiers().stream().anyMatch(annotation -> annotation.annotationType().equals(HealthGroup.class))) {
-                groupHealthChecks.add((HealthCheck) beanManager.getReference(bean, bean.getBeanClass(),
+                groupHealthChecks.add((T) beanManager.getReference(bean, bean.getBeanClass(),
                         beanManager.createCreationalContext(bean)));
             }
         }
 
-        return getHealth(groupHealthChecks);
+        return groupHealthChecks;
     }
 
-    @SafeVarargs
-    private final SmallRyeHealth getHealth(Iterable<HealthCheck>... checks) {
-        JsonArrayBuilder results = jsonProvider.createArrayBuilder();
-        HealthCheckResponse.State status = HealthCheckResponse.State.UP;
+    private SmallRyeHealth getHealth(HealthType... types) {
+        return getHealthAsync(types).await().indefinitely();
+    }
+
+    private Uni<SmallRyeHealth> getHealthAsync(HealthType... types) {
+        List<Iterable<HealthCheck>> checks = new ArrayList<>();
```

Comment:
You can have a single list.

Reply:
well, I guess I can preprocess CDI instances to get only one list of Unis going downwards.